### ISSUE TITLE
Use upgrade-insecure-requests to resolve MCB issues

### ISSUE THREAD

**Issue by gloomy-ghost** [EFForg/https-everywhere#8506]
From https://github.com/EFForg/https-everywhere/pull/7435#issuecomment-279627950:

> Hey guys, @ivysrono just talked to me about [UpgradeMixedContent](https://addons.mozilla.org/zh-CN/android/addon/upgrademixedcontent/). It's an extension that inserts upgrade-insecure-requests to every https page so browsers can rewrite http requests before start mixed content blocking.
> 
> The flaw is inserting such a header let browsers also rewrite display content while it's not blocked by default. If the display content isn't available for HTTPS, it fails. It's quite aggressive to do so for normal users who think http images don't really hurt, but it might be a good way for http-nowhere mode. Any thoughts?

Furthermore, we might be able to do something on mixedcontent rulesets.

**Comment by gloomy-ghost** [comment]
Ping @Hainish

**Comment by jeremyn** [comment]
For reference, [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/upgrade-insecure-requests) is a link describing the `upgrade-insecure-requests` CSP.

**Comment by Hainish** [comment]
It makes sense for us to set the `upgrade-insecure-requests` directive for trivial rulesets.  However, if a rule is not trivial setting `upgrade-insecure-requests` can interfere with *passive* mixed content from loading, since this is not blocked by the major browsers (yet).

**Comment by Hainish** [comment]
Actually it seems like third-party requests are also upgraded by this directive, so we can't be sure that resources will not be blocked *even for trivial rulesets*.

**Comment by Hainish** [comment]
Ah, I misread the initial comment.  Yes, I agree - in HTTP Nowhere mode this should be fine.

**Comment by galeksandrp** [comment]
```xml
<target host="such-a-mixedcontent.host.com" option="upgrade-insecure-requests">
```

Similar #3343

**Comment by iki** [comment]
+1 for adding this, it's badly needed, Many sites like jsfiddle are broken with HTTPS Everywhere, esp. when embedded and the user can't rewrite the resources used easily. 

At the same time, we can't break anything by adding it. It's true that in some cases, the resources can't be loaded by https, so there's no safe solution and the only workaround is to let the user swith to insecure http version of the page easily. These are quite rare for me, I still didn't encounter any, and I hit and solve the MCB issue with HTTPS Everywhere manually quite often. However, while the fix does not work for this corner case, it at least does no damage here, because the https page would not be able to load the http resource due to MCB error anyway.

**Comment by strugee** [comment]
@iki if you find that sites are broken, that's a bug in the ruleset. You should report those.

**Comment by jeremyn** [comment]
@iki `upgrade-insecure-requests` is not safe generally for the reasons @Hainish describes in https://github.com/EFForg/https-everywhere/issues/8506#issuecomment-281478362 and following.

As an example, suppose a request to `https://a.example.com` loads an image (or other passive content) from `http://b.example.com`. Also suppose that `b.example.com` doesn't support HTTPS. If we insert `upgrade-insecure-requests` into the response from `https://a.example.com`, then the image request to `b.example.com` will be changed to HTTPS and will fail. The user then won't see the image that they otherwise would have seen.

**Comment by jeremyn** [comment]
@strugee An active mixed content request is blocked by the browser before HTTPS Everywhere can rewrite the URL (see [the FAQ](https://www.eff.org/https-everywhere/rulesets#mixed-content-blocking-mcb)). `upgrade-insecure-requests` is the way around that problem because it makes the browser rewrite the URL it blocks the request.

**Comment by strugee** [comment]
@jeremyn if things are breaking because of MCB the ruleset should have `platform=mixedcontent` (I've forgotten the exact string but I think you know what I mean).

**Comment by jeremyn** [comment]
@strugee I don't know what's happening with JSFiddle, hopefully @iki can describe what they are seeing in more detail. I agree that adding `platform` is one way to go, although that means it is only active by default on Tor so it's effectively the same as disabling the ruleset by default. We generally let minor active mixed content problems go without marking the ruleset with `platform`. See issue https://github.com/EFForg/https-everywhere/issues/6629 for a related discussion.

**Comment by strugee** [comment]
@jeremyn right, that's why I wanted @iki to report a bug instead of just working around it with upgrade-insecure-requests or whatever ;)

**Comment by Bisaloo** [comment]
Just a reminder that this issue is about adding this header in HTTP-Nowhere mode so it wouldn't break more things than it currently is. In the example described by @jeremyn 

> As an example, suppose a request to https://a.example.com loads an image (or other passive content) from http://b.example.com. Also suppose that b.example.com doesn't support HTTPS. If we insert upgrade-insecure-requests into the response from https://a.example.com, then the image request to b.example.com will be changed to HTTPS and will fail. The user then won't see the image that they otherwise would have seen.

The image will already fail to load in HTTP-Nowhere mode.

**Comment by Bisaloo** [comment]
My issue with this is that an additional header increase indentifiability. This would basically allow websites to find out whether users are using HTTP-Nowhere mode or not.

This could potentially be a problem in Tor browser.

**Comment by jeremyn** [comment]
There is no serious defense against browser fingerprinting for an ordinary non-Tor user, so that shouldn't be a concern.

For Tor, the advice is to use all defaults partly to fight against fingerprinting, so if all Tor users send the new header, then they are no worse off than without the new header.

**Comment by numismatika** [comment]
Is there any progress on the basic firefox issue that causes mcb warnings on ressources we already converted to use tls?

**Comment by jeremyn** [comment]
@numismatika What's an example?

**Comment by numismatika** [comment]
https://forums.askmrrobot.com/ warns about insecure requests from
http://askmrrobot.discoursehosting.net/uploads/db4666/original/1X/80949fd802e399f623154ccd2b33d33a93461cce.png although it is already rewritten by the AMR rule

**Comment by jeremyn** [comment]
@numismatika Thanks, I understand your example. This looks like a Firefox problem, not an HTTPS Everywhere problem. Do you think so too?

**Comment by strugee** [comment]
@jeremyn I believe @numismatika was saying it _was_ an upstream issue and was wondering whether there's been progress on it.

**Comment by strugee** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=878890 appears to be the relevant BMO bug.

**Comment by zero77** [comment]
upgrade-insecure-requests is used by "Smart HTTPS" and it seams to be working very well without there being any problems.

**Comment by ivysrono** [comment]
@zero77 No blacklist on Smart HTTPS.

**Comment by zero77** [comment]
@ivysrono 
There is both a black and white list in Smart HTTPS although, not in HTTPS Everywhere is that what your meaning.

**Comment by Bisaloo** [comment]
`upgrade-insecure-requests` **will** cause issues in cases where a resource is included from a domain that doesn't support `https`.

**Comment by zero77** [comment]
@Bisaloo 
Why not add a time out or look for errors and then downgrade to http.

That way if https is not supported any failed https connection attempts will automatically be downgraded, users will only see a delay and nothing else.

**Comment by ivysrono** [comment]
@zero77 In Smart HTTPS, no blacklist for upgrade-insecure-requests to fix issues in cases where a resource is included from a domain that doesn't support https.

**Comment by strugee** [comment]
@zero77 https://www.eff.org/https-everywhere/faq#why-use-a-whitelist-of-sites-that-support-https-why-cant-you-try-to-use-https-for-every-last-site-and-only-fall-back-to-http-if-it-isnt-available

**Comment by jeremyn** [comment]
@zero77 For your comment https://github.com/EFForg/https-everywhere/issues/8506#issuecomment-333825790, particularly

> Why not add a time out or look for errors and then downgrade to http.

you might find the discussion in issue https://github.com/EFForg/https-everywhere/issues/8239 `Better network errors when "block unencrypted requests" is enabled` relevant, since it talks about challenges related to HTTPS Everywhere detecting and making decisions about connection results.

**Comment by zero77** [comment]
@jeremyn @strugee 
Thanks

**Comment by justyntemme** [comment]
Any progress on this? If help is needed I am more than happy to contribute some time.

**Comment by jeremyn** [comment]
@justyntemme As far as I know there is no active work on this. We seem agreed that adding upgrade-insecure-requests is fine in `Block all unencrypted requests` mode, aside from @Bisaloo 's concerns in https://github.com/EFForg/https-everywhere/issues/8506#issuecomment-321814008 about making a user's browser fingerprint more unique, which I argued was not an issue in https://github.com/EFForg/https-everywhere/issues/8506#issuecomment-321840891.

If you want to create a pull request adding the functionality I described, go ahead, as long as you understand that there's no guarantee it will be merged. There's almost no chance it will be added unless someone does the work first -- @Hainish / EFF staff need to approve it, and they are almost certainly not going to write it themselves or approve it in advance.
